pkg/ipcache/metadata.go

```diff
@@ -292,28 +325,85 @@ func RemoveLabelsFromIPs(
+		toReplace = make(map[string]idPair)
 	)
+
+	IPIdentityCache.Lock()
+	defer IPIdentityCache.Unlock()
+
 	for prefix, lbls := range m {
-		id, exists := IPIdentityCache.LookupByIP(prefix)
+		id, exists := IPIdentityCache.LookupByIPRLocked(prefix)
 		if !exists {
 			continue
 		}
+
 		// Insert to propagate the updated set of labels after removal.
-		var la labels.LabelArray
-		if l := RemoveLabels(prefix, lbls, src); l != nil {
+		//
+		// There are two concurring cases here and they are annotated in the
+		// code below:
+		//   1) When labels are disassociated with an IP, the remaining set
+		//      either contains some labels or it's empty. In the former, the
+		//      identity is both marked for deletion and addition, because
+		//      there are labels still associated with the IP. In the latter,
+		//      the identity is marked for deletion only.
+		//   2) If removeLabels() indicates that a new identity needs to be
+		//      generated (newIdentity), a new identity is allocated (with the
+		//      appropriate labels removed), marked for addition, and lastly,
+		//      upserted into the ipcache (replacing the existing identity
+		//      entry). This can happen when an existing identity was *not*
+		//      released during removeLabels() due to the identity's refcount
+		//      being >1.
+		var (
+			newIdentity bool
+			la          labels.LabelArray
+		)
+		l, newIdentity := removeLabels(prefix, lbls, src)
+		if l != nil {
 			la = l.LabelArray()
 		}
 		idsToDelete[id.ID] = la
-		if len(lbls) > 0 {
+		if len(la) > 0 { // (1)
 			// If for example kube-apiserver label is removed from
-			// a remote-node, then RemoveLabels() will return a
+			// a remote-node, then removeLabels() will return a
 			// non-empty set representing the new full set of
 			// labels to associate with the node. In order to
 			// propagate the new identity, we must emit a delete
 			// event for the old identity and then an add event for
 			// the new identity.
 			idsToAdd[id.ID] = la
 		}
+		if newIdentity { // (2)
+			// If for example kube-apiserver label is removed from a CIDR
+			// identity (contains CIDR labels), then removeLabels() will return
+			// a non-empty set representing just the CIDR labels. We do the
+			// same as (1) from above, but additionally, we must also mark the
+			// new identity for addition and upsert the new identity into the
+			// IPCache.
+			newID, _, err := IdentityAllocator.AllocateIdentity(context.TODO(), l, false)
+			if err != nil {
+				log.WithError(err).WithFields(logrus.Fields{
+					logfields.IPAddr:         prefix,
+					logfields.Identity:       id,
+					logfields.IdentityLabels: l,    // new labels
+					logfields.Labels:         lbls, // removed labels
+				}).Error(
+					"Failed to allocate new identity after dissociating labels from existing identity. Traffic may be disrupted.",
+				)
+				continue
+			}
+			idsToAdd[newID.ID] = la
+			toReplace[prefix] = idPair{
```

Comment:
Isn't `toReplace` effectively just `toAdd` but with a different structure? Can we combine the two?

I note that the current code asks for the old+new to be paired together, but I don't know if that's necessary. If it is, then maybe we should track all of the identities together rather than having three different slices, `toDelete` / `toAdd` / `toReplace`. If not, then maybe we can stick to `toDelete` + `toAdd`?

Reply:
That suggestion makes sense. I was using `toReplace` to combine both operations, so it's pretty much the same thing.

Reply:
I've decided to keep `toReplace` given that this will be the call site that does a force upsert into the ipcache. `toReplace` will help make this operaton explicit, in a way.

Reply:
:+1: although it's effectively similar info to `idsToAdd` / `idsToDelete`, the format of the information is different and it's probably more effort than it's worth to merge these all together, particularly given that these other two must be in a particular type of structure to interact with `UpdateIdentities()`. LGTM.